### PR TITLE
ROX-18480: handler implementation

### PR DESCRIPTION
**Pull request by stehessel**
## Description

Implement the notification handler layer and hook into Central based on the feature flag.

The handler is responsible for

* consuming notifications from the global stream and passing them to the datastore.
* flushing the datastore periodically to initiate the batched write to the database.
* (NOT IMPLEMENTED YET / coming soon) triggering garbage collection of outdated events (lifetime ~3 days or so). This will be done in the context of [ROX-18486](https://issues.redhat.com/browse/ROX-18486).

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

Tested the API with a local Central:
```sh
export MAIN_IMAGE_TAG=4.2.x-9-ga7a588be95
export OUTPUT_FORMAT=helm
export ROX_ADMINISTRATION_EVENTS=true

# Deploy Central to cluster of your choosing.
./deploy/openshift/deploy.sh

# Log in to Central and create an API token.
export ROX_API_TOKEN="ey..."

roxcurl "/v1/administration/events"
```

- see added unit tests.

### CODE REVIEW

File: central/notifications/handler/handler.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+)
+
+var (
+	_ Handler = (*handlerImpl)(nil)
+
+	notificationWriteCtx = sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS, storage.Access_READ_WRITE_ACCESS),
```
Comment (dhaus67) [inline central/notifications/handler/handler.go:21]: Technically, only `READ_WRITE_ACCESS` would suffice here, right?

File: central/notifications/handler/handler.go [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+func (h *handlerImpl) runDatastoreFlush() {
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := h.ds.Flush(notificationWriteCtx); err != nil {
+				log.Error("failed to flush notifications")
```
Comment (dhaus67) [inline central/notifications/handler/handler.go:73]: Would be nice to log the error as well here.

File: central/notifications/handler/handler_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+	s.datastore = dsMocks.NewMockDataStore(s.mockCtrl)
+	s.notificationStream = notifications.NewStream()
+	s.handler = newHandler(s.datastore, s.notificationStream).(*handlerImpl)
+	flushInterval = 10 * time.Millisecond
+	s.handler.Start()
+}
+
+func (s *handlerTestSuite) TearDownTest() {
+	s.mockCtrl.Finish()
```
Comment (dhaus67) [inline central/notifications/handler/handler_test.go:42]: Same comment as in other PRs w.r.t [gomock.Controller.Finish](https://pkg.go.dev/github.com/golang/mock/gomock#Controller.Finish) not being required

File: central/administration/events/handler/singleton.go [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+var (
+	once sync.Once
+
+	h Handler
+)
+
+// Singleton returns an instance of the administration events handler.
+func Singleton() Handler {
+	if !features.AdministrationEvents.Enabled() {
```
Comment (connorgorman) [inline central/administration/events/handler/singleton.go:18]: same nit as other PRs, move this out

File: central/administration/events/handler/handler.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+)
+
+var (
+	_ Handler = (*handlerImpl)(nil)
+
+	flushInterval = time.Minute
```
Comment (connorgorman) [inline central/administration/events/handler/handler.go:19]: nit: add an env var as a safety net to change this with the default as 1 minute. You never know if you need to tune it and when it's hardcoded, it's a nightmare code fix

File: pkg/administration/events/queue.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
+package events
```
Comment (connorgorman) [inline pkg/administration/events/queue.go:1]: Would be good to add a metric to this so that we can track queue size?

**Review by connorgorman — APPROVED** [review]
just one comment on adding some metrics so we can see queue size or we could track queue drops

File: pkg/administration/events/queue.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"container/list"
+
+	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
+)
+
+const (
+	maxQueueSize = 100000
```
Comment (connorgorman) [inline pkg/administration/events/queue.go:11]: do we have any back of envelope math on size of each message so if we hit 100k we think we'd be okay memory wise?

File: pkg/administration/events/queue.go (hunk 6 above)
Comment (dhaus67) [reply pkg/administration/events/queue.go]: Had to do a bit of refactoring to allow metrics being included due to cyclic import: now the `logging` package only relies on the interface and doesn't call `Singleton` anymore (this is anyways what I wanted to in a follow-up).

File: pkg/administration/events/queue.go (hunk 7 above)
Comment (dhaus67) [reply pkg/administration/events/queue.go]: Added a comment, 100k events should be around 16MB in memory.

File: central/administration/events/handler/handler.go [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+	ticker := time.NewTicker(flushInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := h.ds.Flush(h.eventWriteCtx); err != nil {
+				log.Error(err)
+			}
+		case <-h.stopSignal.Done():
```
Comment (rukletsov) [inline central/administration/events/handler/handler.go:78]: Don't we want to flush one last time before stopping?

File: pkg/administration/events/streams/stream.go [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+// GetStreamForTesting creates a new stream for testing purposes.
+func GetStreamForTesting(_ *testing.T) *streamImpl {
+	return newStream()
+}
+
+type streamImpl struct {
+	queue *administrationEventsQueue
+}
+
+// Consume returns the channel to retrieve administration events.
```
Comment (rukletsov) [inline pkg/administration/events/streams/stream.go:26]: What?

File: pkg/administration/events/streams/stream.go [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+func (s *streamImpl) Consume(waitable concurrency.Waitable) *events.AdministrationEvent {
+	return s.queue.pullBlocking(waitable)
+}
+
+// Produce adds an event to the stream.
+//
+// Should be retried with `retry.WithRetry(s.Produce(event))`.
```
Comment (rukletsov) [inline pkg/administration/events/streams/stream.go:33]: Is this recommendation still valid?

File: pkg/administration/events/streams/queue.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	}
+	return event
+}
+
+func (q *administrationEventsQueue) push(event *events.AdministrationEvent) {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	if q.queue.Len() >= maxQueueSize {
+		return
```
Comment (rukletsov) [inline pkg/administration/events/streams/queue.go:77]: Shall we log this on warning level (without creating an _event_, of course)

**Review by rukletsov — APPROVED** [review]
merge time modulo comments

File: pkg/administration/events/streams/queue.go (hunk 11 above)
Comment (dhaus67) [reply pkg/administration/events/streams/queue.go:77]: yeah now we can, initially this wasn't in a separate package and lead to cyclic imports.

**Pull request merged**